ApartmentBuildingSim.py

```diff
@@ -0,0 +1,141 @@
+import heapq
+import random
+import os
+import jsoneng
+from dotenv import load_dotenv
+from termcolor import *
+
+jdb = jsoneng.JsonDB()
+jdb.create({})
+
+
+
+
+
+
+
+
+
+
+class Char:
+    def __init__(self, name):
+        self.name = name
+        self.state = None
+        self.location = None
+        self.pissCounter = 0
+
+    def __repr__(self):
+        return f"CHAR: name:{self.name} state:{self.state} location:{self.location}"
+        # return f"{colored('CHAR:','red')} name:{self.name} state:{self.state} location:{self.location}"
+
+class WaterDrinker(Char):
+    def createEvents(self, simTime, futureEventList, chars):
+        if not any(self in evt.chars for evt in futureEventList):
+            if self.state == 'idle':
+                heapq.heappush(futureEventList, DrinkWater0(simTime + random.randint(1,1), [self]))
+            
+            if self.state == 'need to piss':
+                heapq.heappush(futureEventList, PissWater0(simTime + random.randint(1,1), [self]))
+                self.pissCounter += 1
+
+
+
+
+
+
+
+
+
+
+
+class Event:
+    def __init__(self, time, chars, name):
+        self.time = time
+        self.name = name
+        self.chars = chars
+
+    def __repr__(self):
+        return f"EVENT: time:{self.time} chars:{self.chars[0].name} name:{self.name}"
+
+    def __lt__(self, other):
+        return self.time < other.time
+    
+class InitCharsEvent(Event):
+    def __init__(self, time, chars):
+        super().__init__(time, chars, 'init chars')
+
+    def handleEvent(self, futureEventList):
+        for c in self.chars:
+            c.state = 'idle'
+            c.location = 'lobby'
+    
+class DrinkWater0(Event):
+    def __init__(self, time, chars):
+        super().__init__(time, chars, 'stage 0: start drinking water')
+
+    def handleEvent(self, futureEventList):
+        for c in self.chars:
+            c.state = 'stage 0: drink water'
+        heapq.heappush(futureEventList, DrinkWater1(self.time + random.randint(4,8), self.chars))
+
+class DrinkWater1(Event):
+    def __init__(self, time, chars):
+        super().__init__(time, chars, 'stage 1: finish drinking water')
+
+    def handleEvent(self, futureEventList):
+        for c in self.chars:
+            c.state = 'need to piss'
+
+class PissWater0(Event):
+    def __init__(self, time, chars):
+        super().__init__(time, chars, 'stage 0: start pissing')
+
+    def handleEvent(self, futureEventList):
+        for c in self.chars:
+            c.state = 'stage 0: pissing'
+        heapq.heappush(futureEventList, PissWater1(self.time + random.randint(4,8), self.chars))
+
+class PissWater1(Event):
+    def __init__(self, time, chars):
+        super().__init__(time, chars, 'stage 1: finish pissing')
+
+    def handleEvent(self, futureEventList):
+        for c in self.chars:
+            c.state = 'idle'
```

Comment:
Lots of repeat looping `for c in self.chars`. 
Suggestion: add method `setCharacters(state, action) in class Event
```
class Event:
   ...
   def setCharacters(self, state=None, action=None):
      for c in self.chars:
         if state:
            c.state = state
         if action:
            c.action = action

...

class InitCharsEvent
   def handleEvent:
      self.setCharacters('idle', 'lobby')
 
```